### DISCUSSION TITLE
Running ./hammerdbcli auto <tcl-file> in AWS Lambda function

### DISCUSSION THREAD

**Discussion by joshuajin** [TPC-Council/HammerDB#319] [Q&A] [answered]
I was trying to generate a HammerDB workload from a AWS Lambda function. Specifically running a tcl file to build a schema on a SQL Server database. I tested the ./hammerdbcli auto tcl-file.tcl in a Docker container, that works fine. However after building a AWL Lambda layer with a custom runtime (boostrap), the Lambda function executed ./hammerdbcli -help successfully, but got following error. Has anyone seen these before?

{ [49 bytes data]
* Connection [#0](https://amzn-aws.slack.com/archives/C018BNG2TTK) to host 127.0.0.1 left intact
While loading module "tclreadline"...
no such variable
(read trace on "::env(HOME)")
invoked from within
"variable HISTFILE $::env(HOME)/.tclline_history"
(in namespace eval "::TclReadLine" script line 34)
invoked from within
"namespace eval TclReadLine {
namespace export interact
# Initialise our own env variables:
variable PROMPT "hammerdb>"
variable COMP..."
(file "./modules/tclreadline-1.2.tm" line 28)
invoked from within
"source -encoding utf-8 ./modules/tclreadline-1.2.tm"
("package ifneeded tclreadline 1.2" script)
invoked from within
"package require $m "
END RequestId: a6af8e66-8c1e-44ad-bf67-819228eb1427
REPORT RequestId: a6af8e66-8c1e-44ad-bf67-819228eb1427 Duration: 900.60 ms Billed Duration: 955 ms Memory Size: 128 MB Max Memory Used: 29 MB Init Duration: 54.05 ms 
RequestId: a6af8e66-8c1e-44ad-bf67-819228eb1427 Error: Runtime exited with error: exit status 1
Runtime.ExitErrorRequest ID

**Comment by sm-shaw** [comment, marked as answer]
I've not seen these before, however it looks like the issue is that the system does not have a home environment variable set. This is expected on both Linux and Windows e.g. 

Linux
```
$echo $HOME
/home/steve
$ tclsh
% puts $::env(HOME)
/home/steve
```
Windows
```
C:\Users\Steve>tclsh
% puts $::env(HOME)
C:\Users\Steve
%
```
It will then create a .tclline_history file in this home location to store the previous commands for recall. 

As a first step either see if a HOME environment variable can be set to a writable location or look in modules directory for the file tclreadline-1.2.tm and comment out where it uses $::env(HOME)

```
#variable HISTFILE $::env(HOME)/.tclline_history
#variable  RCFILE $::env(HOME)/.tcllinerc
```
This will cause interactive mode to fail for now but may work for auto. There is also $::env(PATH) used later on so this could possibly fail as well. 
If removing $::env(HOME) works then this could be a solution as the command history is not actually needed when not running in interactive mode.

**Comment by joshuajin** [comment]
Yes, by commenting out the following two lines, now it works in my AWS Lambda function:  ./hammerdbcli auto my-tcl-script.tcl 
#variable HISTFILE $::env(HOME)/.tclline_history
#variable  RCFILE $::env(HOME)/.tcllinerc

Thanks Steve.

**Comment by sm-shaw** [comment]
Great, I will convert this into an issue and provide an adaptable fix in the next version. The history file is not needed if not running in interactive mode. Also, the interactive mode should be able to run if the history file is not found and just have no history of commands to recall.  This will prevent anyone seeing the same error in future.  The planned fix will check to see if the environment variables exist instead of expecting them to be there.
